packages/node/src/indexer/api.service.ts

```diff
@@ -26,15 +29,69 @@ const NOT_SUPPORT = (name: string) => () => {
 
 // https://github.com/polkadot-js/api/blob/12750bc83d8d7f01957896a80a7ba948ba3690b7/packages/rpc-provider/src/ws/index.ts#L43
 const RETRY_DELAY = 2_500;
+const RECONNECT_INTERVAL = 60000;
 
 const logger = getLogger('api');
 
+export class ApiResponseTimeBuffer extends Queue<number> {
+  constructor(capacity: number) {
+    super(capacity);
+  }
+
+  average() {
+    if (this.items.length === 0) {
+      return 0;
+    }
+    return this.items.reduce((a, b) => a + b, 0) / this.items.length;
+  }
+}
+
+export class ApiLoadBalancer {
+  private buffers: ApiResponseTimeBuffer[] = [];
+
+  constructor(numberOfConnections: number, batchCapacity: number) {
+    // fill buffers with ApiResponseTimeBuffer object with capacity of batchCapacity
+    range(0, numberOfConnections).forEach((n) => {
+      this.buffers.push(new ApiResponseTimeBuffer(batchCapacity));
+    });
+    logger.info(`length: ${this.buffers.length}`);
+  }
+
+  addToBuffer(connectionIndex: number, responseTime: number) {
+    if (
+      this.buffers[connectionIndex].size ===
+      this.buffers[connectionIndex].capacity
+    ) {
+      this.buffers[connectionIndex].take();
+    }
+    this.buffers[connectionIndex].put(responseTime);
+  }
+
+  getWeights(disconnectedIndices?: number[]): number[] {
+    const weights = this.buffers.map((buffer, index) => {
+      if (disconnectedIndices?.includes(index)) {
+        return 0;
+      }
+      return buffer.average();
+    });
+    const total = weights.reduce((a, b) => a + b, 0);
+    //deal with the case where average is 0
+    if (total === 0) {
+      return weights.map((weight) => 1 / this.buffers.length);
+    }
+    return weights.map((weight) => weight / total);
+  }
+}
+
 @Injectable()
 export class ApiService implements OnApplicationShutdown {
```

Comment:
I think we can simplify this a lot and remove any external api changes. Externally the api service should not show there are multiple endpoints.

Some things to change:

- Have a connection pool that adds instances when they connect and removes when they disconnect
- We don't need `attemptReconnects` to be on an interval. We can remove an interval and retry as soon as it disconnects
- Have a getter for the api that returns the right instance each time. This might mean updating other classes to use `apiService.api` rather than keeping an internal reference.
- I think we can move `fetchBlockBatches` from being a static function to a class member of api.service. We have done this with etheruem and its working well. That way we can better monitor the requests.
- Retry logic if requests fail should live in the api service.